[WIP] Split shopinvader in 2 modules base and product

matthieusaison opened the pull request:
Resolves https://github.com/shopinvader/odoo-shopinvader/issues/1372

Commit:
Split shopinvader in 2 modules

Commit:
Fix v1_base test

Commit:
Fix install shopinvader_v1_product

Commit:
Hold - fix xmlid and dependency name for shopinvader -> shopinvader_v1_base

lmignon commented:
Hi @matthieusaison 

All this seems to be a hard work. Do you plan to finalize this work in the coming days or do you need some helo on it?

Regards,

lmi

sebastienbeau commented on shopinvader_multi_category/__manifest__.py:
```diff
@@ -10,7 +10,7 @@
     "author": "Akretion",
     "license": "AGPL-3",
-    "depends": ["shopinvader", "product_multi_category"],
+    "depends": ["shopinvader_v1_base", "product_multi_category"],
```
```suggestion
    "depends": ["shopinvader_v1_product", "product_multi_category"],
```

sebastienbeau commented on shopinvader_product_media/__manifest__.py:
```diff
@@ -12,6 +12,6 @@
     "installable": True,
     "auto_install": True,
-    "depends": ["shopinvader", "storage_media_product"],
+    "depends": ["shopinvader_v1_base", "storage_media_product"],
```
```suggestion
    "depends": ["shopinvader_v1_product", "storage_media_product"],
```

sebastienbeau commented on shopinvader_search_engine/__manifest__.py:
```diff
@@ -12,7 +12,7 @@
     "category": "Generic Modules",
     "depends": [
-        "shopinvader",
+        "shopinvader_v1_base",
```
```suggestion
        "shopinvader_v1_product",
```

sebastienbeau commented on shopinvader_v1_base/migrations/14.0.5.14.1/post-migrate.py:
```diff
@@ -9,8 +9,8 @@ def _upgrade_shopinvader_model_access(env):
     Upgrade the "shopinvader partner binding edit" access rights
```
TODO
- not sure that the migration script should be moved, if you have the module installed, I thinks it should stay in shopinvader module

sebastienbeau commented on shopinvader_v1_base/models/shopinvader_partner.py:
```diff
@@ -51,7 +53,7 @@ def _is_shopinvader_user_active(self):
         (
             "record_uniq",
-            "unique(backend_id, record_id, partner_email)",
+            "unique(sync_date, record_id, partner_email)",
```
```suggestion
            "unique(backend_id, record_id, partner_email)",
```

sebastienbeau commented on shopinvader_v1_base/services/abstract_sale.py:
```diff
@@ -43,27 +43,34 @@ def _convert_step(self, sale):
-                self.shopinvader_backend, line.order_id.partner_id.lang
-            )
-        product = self._convert_one_line_product(variant)
+        # variant = line.product_id._get_invader_variant(
```
TODO finish

sebastienbeau commented on shopinvader_v1_base/services/cart.py:
```diff
@@ -289,15 +289,8 @@ def _clear_cart(self, cart):
-
     def _add_item(self, cart, params):
-        self._check_allowed_product(cart, params)
+        # self._check_allowed_product(cart, params) # MOVED IN PRODUCT
```
TODO finish

sebastienbeau commented on shopinvader_v1_base/tests/common.py:
```diff
@@ -93,8 +95,8 @@ class CommonMixin(RegistryMixin, ComponentMixin, UtilsMixin):
-        cls.backend = cls.env.ref("shopinvader.backend_1")
-        cls.backend.bind_all_product()
+        cls.backend = cls.env.ref("shopinvader_v1_base.backend_1")
+        # cls.backend.bind_all_product()
```
TODO remove the line

sebastienbeau commented on shopinvader_v1_product/models/shopinvader_variant.py:
```diff
@@ -9,8 +9,8 @@
 
-from ..utils import float_round
-from .tools import sanitize_attr_name
+from odoo.addons.shopinvader_v1_base.models.tools import sanitize_attr_name
```
TODO check if some tools should be moved into shopinvader_v1_product

sebastienbeau commented on shopinvader_v1_product/services/cart.py:
```diff
@@ -0,0 +1,24 @@
+
+    _inherit = "shopinvader.cart.service"
+
+    def _check_allowed_product(self, cart, params):
```
TODO call this method on add item

sebastienbeau commented on shopinvader_v1_product/tests/test_backend.py:
```diff
@@ -94,7 +92,7 @@ def test_bind_all_product_and_autobind_category_root_level(self):
         self.backend.category_binding_level = 3
         records = self._setup_bind_all_products()
-        self._bind_all_product([("id", "in", records["all_prods"].ids)])
+        # self._bind_all_product([("id", "in", records["all_prods"].ids)])
```
TODO check

sebastienbeau commented on shopinvader_v1_product/tests/test_backend.py:
```diff
@@ -11,12 +11,6 @@ class BackendCase(CommonCase):
     @classmethod
     def setUpClass(cls):
         super(BackendCase, cls).setUpClass()
```
A lot of file here import the file common, but the file is missing, maybe you forget to commit it

Commit:
Changes after review

matthieusaison commented:
Hi @lmignon 
I was in vacation last weeks. I'm back on this work.
It's a tricky work, thanks to propose your help.

Commit:
fix tests

Commit:
split shopinvader_backend

Commit:
update import ref for shopinvader

Commit:
fix ref to shopinvader

sebastienbeau committed:
shopinvader_v1_base: purge registry after test

sebastienbeau committed:
shopinvader_*: finish to adapt split (test, data)

sebastienbeau committed:
shopinvader: clean code for xml compatibility and update xml list

sebastienbeau committed:
shopinvader_sale_profile: fix issue in multi-company

If the partner do not belong to a specific company
the profile is never set
we should use the company of the binding (related backend), not the company of the partner

sebastienbeau commented:
Test are green !! \o/ 
We need to add the migration script and test it with a real project

sebastienbeau commented:
@lmignon you can start to review

lmignon commented:
@matthieusaison @sebastienbeau Thank you for this hard work. All this seems to go into the right way. I haven't found any issue at first read of this giant PR.

sebastienbeau committed:
shopinvader: improve xmlid compatibility

sebastienbeau committed:
Split: Rename module

sebastienbeau committed:
shopinvader_product_binding: remove useless duplicated file

sebastienbeau committed:
shopinvader_restapi, shopinvader_product_binding: add migration script

hparfr commented on shopinvader_product_binding/__manifest__.py:
```diff
@@ -0,0 +1,61 @@
+        "shopinvader_restapi",
+        # "component_event",  # product_event_listener
+        # "base_url",  # shopinvader.category
+        # "base_sparse_field_list_support",  # shopinvader.binding
```
```suggestion
```

these modules are added just few lines after;

matthieusaison replied:
thanks, 
Change was done here https://github.com/shopinvader/odoo-shopinvader/pull/1382/commits/1fd8d89bc8c6e78093c63e8cbb72d04f2ba42efe

qgroulard commented on shopinvader_restapi/models/shopinvader_backend.py:
```diff
@@ -70,28 +64,6 @@ class ShopinvaderBackend(models.Model):
         "field on the sale order created.",
     )
     filter_ids = fields.Many2many(comodel_name="product.filter", string="Filter")
```
`product.filter` is now defined in `shopinvader_product_binding`.

matthieusaison replied:
thanks for the feedback, [53839ec](https://github.com/shopinvader/odoo-shopinvader/pull/1382/commits/53839ec0e3327c407058187016422a6e8f7dfec1)

marielejeune commented on shopinvader_restapi/views/shopinvader_menu.xml:
```diff
@@ -22,13 +22,6 @@
         groups="sales_team.group_sale_salesman"
```
Don't forget to change web_icon path line 13 of this file (sorry but GitHub doesn't let me point out a line with no change):

`web_icon="shopinvader,static/description/icon.png"` -> `web_icon="shopinvader_restapi,static/description/icon.png"`

marielejeune reviewed (commented):
Thanks for this huge work!
Just one minor comment

hparfr commented on shopinvader_product_binding/__manifest__.py:
```diff
@@ -0,0 +1,61 @@
+        "views/shopinvader_product_view.xml",
+        "views/shopinvader_variant_view.xml",
+        "views/shopinvader_category_view.xml",
+        "views/shopinvader_backend_view.xml",
```
why moving shopinvader backend  and shopinvader menu view ?

matthieusaison replied:
The shopinvader menu view has been split, with only the menu linked to the product in the product_binding module.
[https://github.com/akretion/odoo-shopinvader/blob/14.0-split_shopinvader_v1_product/shopinvader_restapi/views/shopinvader_menu.xml](url)

However, the shopinvader backend view hasn't been split as it does, so I'll take care of that.

hparfr replied:
Hi, the more I digg into it, the more I doubt it's the good solution.
~~We need one basic shoinvader module (with basic menus), shopinvader_restapi should depend of it~~

Edit: now I have a better understanding.

- Basic menus entries and security rules: shopinvader
- shopinvader_backend -> shopinvader_restapi
- ir_exports for products, categorie -> shopinvader (or shopinvader_se_product but very light module)

matthieusaison replied:
According to Seb, the menu have to stay in shopinvader_restapisecurity, 
ir_export product and category stays in shopinvader_product_binding

hparfr commented on shopinvader_product_binding/__manifest__.py:
```diff
@@ -0,0 +1,61 @@
+        "demo/product_product_demo.xml",
+    ],
+    "pre_init_hook": "pre_init_hook",
+}
```
can you be more explicit about what the module shopinvader_product_binding should do and its responsibility.

sebastienbeau replied:
maybe something like

On V14 and previous version product was bind to a shopinvader backend. On futur version this logic is remove. Product will be linked to a sale channel and the sale_channel_search_engine will take care of exporting then to the search engine.
Do not migrate this module on V16

matthieusaison replied:
description added

marielejeune commented:
`shopinvader.variant` and `shopinvader.variant.binding.wizard` are now defined in `shopinvader_product_binding` which is not a dependency of `shopinvader_restapi`.
This test class methods must thus be adapted: https://github.com/shopinvader/odoo-shopinvader/blob/e42e6848ed9875d848daf4c49528d2c022e38bb0/shopinvader_restapi/tests/common.py#L53C5-L74C19 and https://github.com/shopinvader/odoo-shopinvader/blob/e42e6848ed9875d848daf4c49528d2c022e38bb0/shopinvader_restapi/tests/common.py#L41C5-L51C36

Commit:
review change

Commit:
split shopinvader backend view

Commit:
move filter_ids

Commit:
adapt test class method for product

marielejeune commented on shopinvader_restapi/services/abstract_sale.py:
```diff
@@ -43,27 +43,14 @@ def _convert_step(self, sale):
         return True
 
+    def _get_product_information(self, line):
+        return {"id": line.product_id.id}
```
Before, due to the shopinvader product variant, we got all info about a product inside the SO data (product name, prices, + various custom fields...). Now I see that we just get the product ID. Is it wanted? What are the benefits?

Commit:
readme

Commit:
fix

matthieusaison commented:
Yes it's wanted, the variant case is now handle
in shopinvader_product_binding, because the class ShopinvaderVariant has
moved here too.
https://github.com/shopinvader/odoo-shopinvader/blob/6d3ed9e88adb1f6e718184dba1fd37506a690de8/shopinvader_product_binding/services/abstract_sale.py#L33

Le lun. 18 sept. 2023 à 17:40, Marie Lejeune ***@***.***> a
écrit :

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In shopinvader_restapi/services/abstract_sale.py
> <https://github.com/shopinvader/odoo-shopinvader/pull/1382#discussion_r1328926014>
> :
>
> > @@ -43,27 +43,14 @@ def _convert_step(self, sale):
>      def _is_item(self, line):
>          return True
>
> +    def _get_product_information(self, line):
> +        return {"id": line.product_id.id}
>
> Before, due to the shopinvader product variant, we got all info about a
> product inside the SO data (product name, prices, + various custom
> fields...). Now I see that we just get the product ID. Is it wanted? What
> are the benefits?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/shopinvader/odoo-shopinvader/pull/1382#pullrequestreview-1631388276>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5CI2NJ6AMSSISR4ILDKDQTX3BTPVANCNFSM6AAAAAAZ4EY2WM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
